content/docs/grammar-language.md

```diff
@@ -0,0 +1,371 @@
+```
+Here a `Person` object **needs** to first declare the property `name` then `age`.
+```
+person Bob 25
+```
+will successfully be parsed to an object of type `Person` while
+```
+person 25 Bob
+```
+will throw an error.
+
+However, it is possible to declare a group of properties in an unordered fashion using the `&` operator
+```
+Person:
+    'person' name=ID & age=INT
+```
+will now allow `name` and `age` to be declared in any order.
+```
+person 25 Bob
+```
+will then successfully create an object of type `Person`.
+
+Cardinality (?,*,+ operators) also applies to unordered group. Please note that assignments with a cardinality of `+` or `*` have to appear continuously and cannot be interrupted by an other assignment and resumed later.
+
+#### Simple Actions
+It is possible for a rule to return different types depending on declaration
+```
+RuleOne returns TypeOne:
+    'keywordOne' name=ID | RuleTwo;
+
+RuleTwo returns TypeTwo:
+    'keywordTwo' name=ID;
+```
+In the above example, we rely on a *rule call* to specify the return type. With more complex rules, the readability will be highly impacted. *Actions* allow to improve the readability of the grammar by explicitly defining the return type. *Actions* are declared inside of curly brackets `{}`:
+```
+RuleName returns TypeOne:
+    'keywordOne' name=ID | 'keywordTwo' {TypeTwo} name=ID;
+```
+
+#### Unassigned Rule Calls
+Parser rules do not necessarily need to return an object, they can also refer to other parser rules which in turn will be responsible for returning the object.
+For example, in the [Arithmetics example](https://github.com/langium/langium/blob/main/examples/arithmetics/src/language-server/arithmetics.langium):
+```
+AbstractDefinition:
+	Definition | DeclaredParameter;
+```
+The parser rule `AbstractDefinition` will not create an object of type AbstractDefinition. Instead, it calls either the `Definition` or `DeclaredParameter` parser rule which will be responsible for creating an object of a given type (or call other parser rules if they are unassigned rule calls themselves). 
+
+#### Assigned Actions
+The parser is built using [Chevrotain](https://github.com/chevrotain/chevrotain) which implements a LL(k) parsing algorithm (left-to-right). By definition, a LL(k) grammar cannot have rules containing left recursion.
+
+Consider the following: 
+```
+Addition:
+    Addition '+' Addition | '(' Addition ')' | value=INT;
+```
+The parser rule `Addition` is left-recursive and will not be parsable. We can go around this issue by *left-factoring* the rule, *i.e.* by factoring out the common left-factor. We introduce a new rule `Expression`:
+```
+Expression:
+    '(' Addition ')' | value=INT;
+
+Addition:
+    Expression ('+' Expression)*;
+```
+
+[NOTE] I left out the part regarding `tree rewrite actions` as I think it could be part of a more advanced documentation part on the AST/CST. I can add it if we think it belongs here.
+
+### Syntactic Predicates
```

Comment:
Syntactic predicates were only needed in Xtext due to limitations of the underlying parser generator (Antlr3). We don't need syntactic predicates in Langium. We support them, but only for backwards compatibility.

Reply:
Should we then mention that it is supported for backwards compatibility but not recommend it for new grammars? Can we then just link to the Xtext documentation?

Reply:
The grammar is backwards compatible in a way that enables syntactic predicates from a grammar parser perspective , but it does not do anything in any way. So I think we shouldn't mention it at all.

Reply:
Removed